### PR TITLE
ds_prefill(MoE plumbing: gate internalized, routing setup, test infrastructure)

### PR DESCRIPTION
**Pull request by mbezuljTT**
### Ticket
Link to GitHub issue

### Problem description
The MoE pipeline required external gate construction and manual routing setup before calling `TtMoe` or `TorchMoe`. Gate weights, indices, offsets, and token counts were all computed externally and passed as arguments, making the API fragile and tightly coupled to the caller.

### What's changed

Both `TtMoe` and `TorchMoe` now create their gate internally from a `gate_weights` dict. `TtMoe.forward(x)` and `TorchMoe.forward(x)` run the full pipeline end-to-end including gate — callers no longer construct gate objects or compute routing tensors.

**Routing setup** — Separated dispatch offset/count computation from the old gate into `TtMoERoutingSetup` (`tt_moe_routing_setup.py`), reusable by both gate and dispatch paths. Removed redundant `cum_sum` from `get_gate_outputs` return value.

**Gate module** — `TtMoEGatePrefill` (`tt_moe_gate_prefill.py`) with `GateFallbackMode` supporting 4 modes: full device, host grouped-gate, host matmul, or full host fallback. Accepts `weight`/`bias` in HF convention `(n_experts, dim)`, transposes internally for TTNN. Replaces old `moe_gate_prefill2d.py`.

**Gate integration** — `TtMoe.__init__` builds gate config + gate + input mem config from init params. `TorchMoe.__init__` builds `ReferenceMoEGate` when `gate_weights` provided, falls back to explicit routing when `gate_weights=None`.

**Dispatch/Combine** — Dispatch+combine test now uses `MoERoutingSetup` to compute offsets and counts. `expert_offsets` shape changed to per-dispatch-group `(num_dispatch_groups, dispatch_group_size, num_routed_experts)` with updated sharding `dims=(1, 0)`. Combine now uses `ExpertMapping.get_global_expert_idx()` to look up token counts with the correct global expert index. Dispatch offset aligned to `uint32_t`.

**Intermediates** — Both `MoEIntermediates` and `TtMoEIntermediates` now expose `gate_scores`, `gate_indices`, `gate_logits`, `expert_token_counts`.

**Test infrastructure** — Gate prefill test refactored with `GateFallbackMode` parametrization and sorted indices for recall. TTNN MoE test updated for integrated gate (using `GateFallbackMode.HOST_ALL` for now) with profiler, gate recall, dense PCC, sparse validation. Shared helpers (`create_gate_weights`, `create_torch_expert_weights`, `create_shared_expert_weights`) moved to `init_helpers.py`.

#### Model tests

- [ ] [blackhole-e2e-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-e2e-tests.yaml?query=branch:mbezulj/2603-moe-plumbing)
- [ ] [t3000-e2e-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-e2e-tests.yaml?query=branch:mbezulj/2603-moe-plumbing)
- [ ] [(Galaxy) DeepSeek Prefill tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-deepseek-prefill-tests.yaml?query=branch:mbezulj/2603-moe-plumbing)
- [ ] [(Release) Model Status for Console Deployment](https://github.com/tenstorrent/tt-metal/actions/workflows/demo-sp-release.yaml?query=branch:mbezulj/2603-moe-plumbing)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/combine/device/combine_device_operation.cpp [hunk 1]
```diff
@@ -54,12 +55,12 @@ void CombineDeviceOperation::validate_on_program_cache_miss(
     TT_FATAL(
         dispatched_shape[0] == metadata_shape[0] && dispatched_shape[0] == counter_shape[0],
         "First dimension (per_device_batch) must match across all input tensors");
-    TT_FATAL(
-        dispatched_shape[2] == metadata_shape[2] && dispatched_shape[2] == counter_shape[2],
-        "experts_per_chip dimension must match: dispatched[2]={}, metadata[2]={}, counter[2]={}",
-        dispatched_shape[2],
-        metadata_shape[2],
-        counter_shape[2]);
+    // TT_FATAL(
```
Comment (mbezuljTT) [inline ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/combine/device/combine_device_operation.cpp]: return and adjust this fatal

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/combine/device/kernels/dataflow/reader_combine.cpp [hunk 2]
```diff
@@ -150,8 +150,16 @@ void kernel_main() {
             i, experts_tok_counter_addr_gen, counter_base_addr + i * aligned_experts_tok_counter_page_size);
     }
     noc_async_read_barrier();
+
+    // Expert token counts are laid out as [n_dispatch_groups, experts_per_dispatch_group]
+    // where each dispatch group is a mesh column (all rows in that column).
+    // Row-major linearization: linearized_mesh_coord = mesh_row * mesh_cols + mesh_col
```
Comment (mbezuljTT) [inline ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/combine/device/kernels/dataflow/reader_combine.cpp:156]: this was key adjustments between routing setup and combine op. @ddjekicTT  @ianastasijevicTT  fyi

File: .github/workflows/t3000-e2e-tests-impl.yaml [hunk 3]
```diff
@@ -106,7 +106,6 @@ jobs:
       - name: ${{ matrix.test-group.name }}
         timeout-minutes: ${{ matrix.test-group.timeout }}
         run: |
-          echo ${{ matrix.test-group.cmd }}
```
Comment (mbezuljTT) [inline .github/workflows/t3000-e2e-tests-impl.yaml:109]: note: this was causing tests to be ran twice because .cmd is multiline string

File: models/demos/deepseek_v3_d_p/tests/pcc/test_moe_routing_setup.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        ),
+        pytest.param(
+            (2, 4),
+            {
+                "fabric_config": ttnn.FabricConfig.FABRIC_1D,
+                "fabric_router_config": create_fabric_router_config(max_payload_size=7 * 1024),
+            },
+            1,
+            ttnn.Topology.Linear,
+            marks=pytest.mark.requires_mesh_topology(mesh_shape=(2, 4), topology="mesh-4x2"),
```
Comment (ddjekicTT) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_moe_routing_setup.py]: nitpick: topology name is "mesh-4x2" - should be "mesh-2x4"

**Review by ddjekicTT — APPROVED** [review]

**Review by dpopovTT — APPROVED** [review]
